renku_notebooks/api/notebooks.py

```diff
@@ -295,11 +301,143 @@ def launch_notebook(
     manifest = server.start()
 
     current_app.logger.debug(f"Server {server.server_name} has been started")
-    for autosave in user.get_autosaves(server.gl_project.path_with_namespace):
-        autosave.cleanup(server.commit_sha)
+
     return NotebookResponse().dump(UserServerManifest(manifest)), 201
 
 
+@bp.route("servers/<server_name>", methods=["PATCH"])
+@use_args(
+    {"server_name": fields.Str(required=True)}, location="view_args", as_kwargs=True
+)
+@use_args(PatchServerRequest(), location="json", as_kwargs=True)
+@authenticated
+def patch_server(user, server_name, state):
+    """
+    Patch a user server by name based on the query param.
+
+    ---
+    patch:
+      description: Patch a running server by name.
+      requestBody:
+        content:
+          application/json:
+            schema: PatchServerRequest
+      parameters:
+        - in: path
+          schema:
+            type: string
+          name: server_name
+          required: true
+          description: The name of the server that should be patched.
+      responses:
+        204:
+          description: The server was hibernated successfully.
+        400:
+          description: Invalid json argument value.
+        404:
+          description: The server cannot be found.
+          content:
+            application/json:
+              schema: ErrorResponse
+        500:
+          description: The server exists but could not be successfully hibernated.
+          content:
+            application/json:
+              schema: ErrorResponse
+      tags:
+        - servers
+    """
+    # TODO: Return an error if the deployment doesn't use PVC for sessions
+
+    logging.warning(f"State is {state}")
+    logging.warning(
+        f"H = {PatchServerStatusEnum.Hibernated.value}, R = {PatchServerStatusEnum.Running.value}"
+    )
+
+    server = config.k8s.client.get_server(server_name, user.safe_username)
+
+    if state == PatchServerStatusEnum.Hibernated.value:
+        # NOTE: Do nothing if server is already hibernated
```

Comment:
The issue with repeated patching is that we can't get repository's status the second time (from the git sidecar). In such cases, we shouldn't overwrite hibernation annotations that's probably retrieved from the sidecar the first time. Therefore, we always need to check if the sessions is hibernated or not.
I can change the code to always patch but don't overwrite hibernation annotations but that's basically the same as not patching.

Resuming is ok and can be repeated with no problem since we remove those annotations.

Reply:
So I think it is easier to do the following on patch:
1. Patch the "hibernated" flag in the spec
2. Attempt to get the status from the sidecar:
  - if you cannot get the status leave the hibernation annotations alone
  - if you can get the status then update the annotations

This way you remove one call to the k8s api on patching. And yes if the session is patched already then essentially this does not do anything. Which is exactly right.